python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py

```diff
@@ -221,6 +227,68 @@ def setup(self, trainer: "pl.Trainer") -> None:
+            optimizers = [_unpack_lightning_optimizer(opt) for opt in optimizers]
+
+            for optimizer in optimizers:
+                for param_group in optimizer.param_groups:
+                    param_group["lr"] *= self.world_size
+
+            lr_scheduler_configs = self.lr_scheduler_configs
+            for config in lr_scheduler_configs:
+                scheduler = config.scheduler
+                scheduler.base_lrs = [lr * self.world_size for lr in scheduler.base_lrs]
+
+            if self.lr_scheduler_configs:
+                warnings.warn(f"Nano warmup currently only support none optimizers, "
+                              f"but got {len(self.lr_scheduler_configs)}. Skip warmup")
+            else:
+                lr_schedulers = []
+                for opt_idx, opt in enumerate(self.optimizers):
+                    if type(self.auto_lr) is bool:
+                        self.auto_lr = trainer.max_epochs // 5 + 1
+                    if not TORCH_VERSION_LESS_1_10:
+                        from torch.optim.lr_scheduler import LinearLR
+                        scheduler = LinearLR(optimizer=opt,
+                                             # set initial lr as user lr
+                                             start_factor=1.0 / self.world_size,
+                                             end_factor=1.0,
+                                             total_iters=self.auto_lr)
+                    else:
+                        from torch.optim.lr_scheduler import LambdaLR
+
+                        def lr_func(epoch):
+                            start_factor = 1.0 / self.world_size
+                            end_factor = 1.0
+                            total_iters = self.auto_lr
+                            if epoch == 0:
+                                return start_factor
+                            if epoch < total_iters:
+                                return (end_factor - start_factor) * epoch / total_iters \
+                                    + start_factor
+                            else:
+                                return 1.0
+                        scheduler = LambdaLR(optimizer=opt,
+                                             lr_lambda=[lr_func] * len(optimizer.param_groups))
+                    lr_scheduler = {
+                        'scheduler': scheduler,
+                        'opt_idx': opt_idx
+                    }
+                    lr_schedulers.append(lr_scheduler)
+                lr_scheduler_configs = (
+                    _configure_schedulers_automatic_opt(lr_schedulers, None)
+                    if self.lightning_module.automatic_optimization
+                    else _configure_schedulers_manual_opt(lr_schedulers)
+                )
+                _set_scheduler_opt_idx(self.optimizers, lr_scheduler_configs)
+                _validate_scheduler_api(lr_scheduler_configs, self.model)
+                self.lr_scheduler_configs = lr_scheduler_configs
```

Comment:
Are these lines adapted from some original code in the pytorch-lightning code base?

Reply:
Maybe add some reference?

Reply:
Yes, they are used to make sure the lr_scheduler are
established correctly

Reply:
https://github.com/Lightning-AI/lightning/blob/1.6.4/pytorch_lightning/core/optimizer.py#L189